### ISSUE TITLE
ConcurrentModificationException in XStreamExecutionContextStringSerializer.serialize when running Partitioned jobs in spring batch

### ISSUE THREAD

**Issue by prakashpalakkarai** [spring-projects/spring-batch#3766]
**Whats happening**

Sporadically I am seeing ConcurrentModificationException in XStreamExecutionContextStringSerializer.serialize when the partitioned job steps are complete and the job is about to complete. From stack trace, I can see this is happening when an attempt is made to serialize JobExectuionContext for updating to database. 

**Exception Stack Trace**

```
[8/18/20 1:30:11:933 EDT] 00011cbd org.springframework.batch.core.step.AbstractStep             E Encountered an error executing step xxxPartitionStep in job xxx
org.springframework.batch.core.JobExecutionException: Flow execution ended unexpectedly
	at org.springframework.batch.core.job.flow.FlowStep.doExecute(FlowStep.java:98)
	at org.springframework.batch.core.step.AbstractStep.execute(AbstractStep.java:200)
	at org.springframework.batch.core.partition.support.TaskExecutorPartitionHandler$1.call(TaskExecutorPartitionHandler.java:139)
	at org.springframework.batch.core.partition.support.TaskExecutorPartitionHandler$1.call(TaskExecutorPartitionHandler.java:136)
	at java.util.concurrent.FutureTask.run(FutureTask.java:277)
	at java.lang.Thread.run(Thread.java:785)
Caused by: org.springframework.batch.core.job.flow.FlowExecutionException: Ended flow=xxxPartitionStep at state=xxxPartitionStep.xxxyyyPendInCrStep with exception
	at org.springframework.batch.core.job.flow.support.SimpleFlow.resume(SimpleFlow.java:178)
	at org.springframework.batch.core.job.flow.support.SimpleFlow.start(SimpleFlow.java:144)
	at org.springframework.batch.core.job.flow.FlowStep.doExecute(FlowStep.java:90)
	... 5 more
Caused by: java.util.ConcurrentModificationException
	at java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:730)
	at java.util.LinkedHashMap$LinkedEntryIterator.next(LinkedHashMap.java:763)
	at java.util.LinkedHashMap$LinkedEntryIterator.next(LinkedHashMap.java:761)
	at com.thoughtworks.xstream.converters.collections.MapConverter.marshal(MapConverter.java:75)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller.convert(AbstractReferenceMarshaller.java:69)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:58)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller$1.convertAnother(AbstractReferenceMarshaller.java:84)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.marshallField(AbstractReflectionConverter.java:274)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter$2.writeField(AbstractReflectionConverter.java:250)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter$2.<init>(AbstractReflectionConverter.java:213)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.doMarshal(AbstractReflectionConverter.java:144)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.marshal(AbstractReflectionConverter.java:90)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller.convert(AbstractReferenceMarshaller.java:69)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:58)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller$1.convertAnother(AbstractReferenceMarshaller.java:84)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.marshallField(AbstractReflectionConverter.java:274)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter$2.writeField(AbstractReflectionConverter.java:250)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter$2.<init>(AbstractReflectionConverter.java:213)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.doMarshal(AbstractReflectionConverter.java:144)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.marshal(AbstractReflectionConverter.java:90)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller.convert(AbstractReferenceMarshaller.java:69)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:58)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:43)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller$1.convertAnother(AbstractReferenceMarshaller.java:88)
	at com.thoughtworks.xstream.converters.collections.AbstractCollectionConverter.writeItem(AbstractCollectionConverter.java:66)
	at com.thoughtworks.xstream.converters.collections.MapConverter.marshal(MapConverter.java:79)
	at com.thoughtworks.xstream.core.AbstractReferenceMarshaller.convert(AbstractReferenceMarshaller.java:69)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:58)
	at com.thoughtworks.xstream.core.TreeMarshaller.convertAnother(TreeMarshaller.java:43)
	at com.thoughtworks.xstream.core.TreeMarshaller.start(TreeMarshaller.java:82)
	at com.thoughtworks.xstream.core.AbstractTreeMarshallingStrategy.marshal(AbstractTreeMarshallingStrategy.java:37)
	at com.thoughtworks.xstream.XStream.marshal(XStream.java:1319)
	at com.thoughtworks.xstream.XStream.marshal(XStream.java:1308)
	at com.thoughtworks.xstream.XStream.toXML(XStream.java:1281)
	at com.thoughtworks.xstream.XStream.toXML(XStream.java:1268)
	at org.springframework.batch.core.repository.dao.XStreamExecutionContextStringSerializer.serialize(XStreamExecutionContextStringSerializer.java:90)
	at org.springframework.batch.core.repository.dao.XStreamExecutionContextStringSerializer.serialize(XStreamExecutionContextStringSerializer.java:45)
	at org.springframework.batch.core.repository.dao.JdbcExecutionContextDao.serializeContext(JdbcExecutionContextDao.java:299)
	at org.springframework.batch.core.repository.dao.JdbcExecutionContextDao.updateExecutionContext(JdbcExecutionContextDao.java:144)
	at org.springframework.batch.core.repository.support.SimpleJobRepository.updateExecutionContext(SimpleJobRepository.java:210)
	at sun.reflect.GeneratedMethodAccessor1137.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:302)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:281)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:208)
	at com.sun.proxy.$Proxy93.updateExecutionContext(Unknown Source)
	at org.springframework.batch.core.job.SimpleStepHandler.handleStep(SimpleStepHandler.java:159)
	at org.springframework.batch.core.job.flow.JobFlowExecutor.executeStep(JobFlowExecutor.java:64)
	at org.springframework.batch.core.job.flow.support.state.StepState.handle(StepState.java:67)
	at org.springframework.batch.core.job.flow.support.SimpleFlow.resume(SimpleFlow.java:169)
	... 7 more
```

**Environment**

spring batch version:- 3.0.7.RELEASE
spring framework version: 4.2.4.RELEASE
spring batch admin version: 1.3.1.RELEASE
spring integration version: 4.2.4.RELEASE
xstream version: 1.4.7
Database: DB2 
Java version: 1.7 compile on 1.8 runtime (IBM JDK)

**Steps to Reproduce**

Run Partitioned job with 10 partitions with few steps with steps defined as a flow. 

**Expected Behavior**

Expected to run without issues.

**Minimal Complete Reproducible example**

I have attached a sample job xml that is similar to the production job definition.  

[sample-partition-job.txt](https://github.com/spring-projects/spring-batch/files/5125682/sample-partition-job.txt)

**Comment by benas** [comment]
@prakashpalakkarai Thank you for reporting this issue. However, Spring Batch 3 is not maintained anymore and the `XStreamExecutionContextStringSerializer` has been deprecated for removal in favor of `Jackson2ExecutionContextStringSerializer`. Can you confirm you still have the issue with Spring Batch 4.3.1 and the `Jackson2ExecutionContextStringSerializer`?

**Comment by prakashpalakkarai** [comment]
Thank you for the response. We also use spring batch admin along with Spring Batch 3. Not sure if that is still available with Spring Batch 4.3.1. I tried to see if we can upgrade to Spring Framework 4.3.9 earlier and was not able to because Spring Batch Admin was not supported. Is there an alternative for Spring Batch Admin with the upgrade to Spring Batch 4.3.1 and Spring framework along with it.

**Comment by benas** [comment]
> Is there an alternative for Spring Batch Admin with the upgrade to Spring Batch 4.3.1 and Spring framework along with it.

Yes, it is [Spring Cloud Data Flow](https://dataflow.spring.io). Spring Batch Admin has been deprecated and archived several years ago in favor of SCDF. You can find a migration guide [here](https://github.com/spring-attic/spring-batch-admin/blob/master/MIGRATION.md).

> Can you confirm you still have the issue with Spring Batch 4.3.1 and the Jackson2ExecutionContextStringSerializer?

You did not confirm that, so I'm closing this issue. If you still have the problem with v4.3.1 and `Jackson2ExecutionContextStringSerializer`, please provide a minimal example or a failing test and we will re-open this issue for investigation.
